WhatsApp/app/src/main/java/com/example/whatsapp/ui/fragments/home/chats/ChatsFragment.kt

```diff
@@ -75,18 +105,209 @@ class ChatsFragment : BaseFragment<FragmentChatsBinding>() {
                 chatsAdapter.submitList(it)
             }.autoDispose()
 
+        viewModel.getSelectionType()
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe {
+                updateActionBarState(it)
+            }
+
         //endregion
     }
     //endregion
 
+    //region ActionMode.Callback
+    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
+        mode.menuInflater.inflate(R.menu.home_chats_context_menu, menu)
+        return true
+    }
+
+    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
+        menu.findItem(R.id.pinChats).setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
+        menu.findItem(R.id.deleteChats).setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
+        menu.findItem(R.id.archiveMessages).setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
+        menu.findItem(R.id.muteNotifications).setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
+        return true
+    }
+
+    private fun disableAllActionBarItems() {
+        actionMode?.menu?.findItem(R.id.pinChats)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.deleteChats)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.muteNotifications)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.archiveMessages)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.exitGroup)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.addChatShortcut)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.viewContact)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.addToContact)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.viewContact)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.exitGroups)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.addToContact)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.groupInfo)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.markUnread)?.isVisible = false
+        actionMode?.menu?.findItem(R.id.selectAll)?.isVisible = false
+    }
+
+    private fun updateActionBarState(selectionType: ConversationSelectionType) {
+        when (selectionType) {
```

Comment:
Is there maybe any menu items where it makes sense that they can be categorised together to reduce the size of these methods

Also is findItem as heavy as findviewbyid? Might be better to initialize these views first or make them lazy if that's the case

Reply:
In this case I don't believe there is as we need them to be in a given order the alternate option to this by using groups means we would have to have duplicate items with the same function like one for groups, one for multiple groups, single dm, multiple dms and mixed results I am not sure if this would end up adding even more code?